arraycontext/parameter_study/transform.py

```diff
@@ -0,0 +1,602 @@
+
+        rec_keyed_map_array_container(_as_dict_of_named_arrays,
+                                      output_template)
+
+        input_shapes = {}
+        input_axes = {}
+        for key, val in arg_id_to_descr.items():
+            if isinstance(val, LeafArrayDescriptor):
+                input_shapes[input_id_to_name_in_program[key]] = val.shape
+            input_axes[input_id_to_name_in_program[key]] = arg_id_to_arg[key].axes
+        expand_map = ExpansionMapper(input_shapes, input_axes)
+        # Get the dependencies
+
+        sing_inst_outs = pt.make_dict_of_named_arrays(dict_of_named_arrays)
+
+        # Use the normal compiler now.
+
+        compiled_func = self._dag_to_compiled_func(expand_map(sing_inst_outs),
+                                                   # pt_dict_of_named_arrays,
+                input_id_to_name_in_program=input_id_to_name_in_program,
+                output_id_to_name_in_program=output_id_to_name_in_program,
+                output_template=output_template)
+
+        self.program_cache[arg_id_to_descr] = compiled_func
+        return compiled_func(arg_id_to_arg)
+
+
+def _cut_if_in_param_study(name, arg) -> Array:
+    """
+        Helper to split a place holder into the base instance shape
+        if it is tagged with a `ParameterStudyAxisTag`
+        to ensure the survival of the information those tags will be converted
+        to temporary Array Tags of the same type. The placeholder will not
+        have the axes marked with a `ParameterStudyAxisTag` tag.
+    """
+    ndim: int = len(arg.shape)
+    newshape = []
+    update_axes: AxesT = (Axis(tags=frozenset()),)
+    for i in range(ndim):
+        axis_tags = arg.axes[i].tags_of_type(ParameterStudyAxisTag)
+        if not axis_tags:
+            update_axes = *update_axes, arg.axes[i],
+            newshape.append(arg.shape[i])
+    # remove the first one that was placed there for typing.
+    update_axes = update_axes[1:]
+    update_tags: FrozenSet[Tag] = arg.tags
+    return pt.make_placeholder(name, newshape, arg.dtype, axes=update_axes,
+                               tags=update_tags)
+
+
+def _get_f_placeholder_args_for_param_study(arg, kw, arg_id_to_name, actx):
+    """
+    Helper for :class:`BaseLazilyCompilingFunctionCaller.__call__`.
+    Returns the placeholder version of an argument to
+    :attr:`ParamStudyLazyPyOpenCLFunctionCaller.f`.
+
+    Note this will modify the shape of the placeholder to
+    remove any parameter study axes until the trace
+    can be completed.
+
+    They will be added back after the trace is complete.
+    """
+    if np.isscalar(arg):
+        name = arg_id_to_name[(kw,)]
+        return pt.make_placeholder(name, (), np.dtype(type(arg)))
+    elif isinstance(arg, pt.Array):
+        name = arg_id_to_name[(kw,)]
+        # Transform the DAG to give metadata inference a chance to do its job
+        arg = _to_input_for_compiled(arg, actx)
+        return _cut_if_in_param_study(name, arg)
```

Comment:
It looks like you first make wrong placeholders and then fix them? As opposed to making correct ones from the get-go? Why?